### PR TITLE
vdpa/virtio: replace doorbell relay with notifier

### PR DESCRIPTION
**Pull request by kailiangz1**
Doorbell relay is used before, but it is time consuming. Use notifier instead and use DPDK work to do notify because we can't wait when we process vhost msg, otherwise, deadlock will happend, so, use work thread to finish the job

### CODE REVIEW

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 1]
```diff
@@ -583,11 +434,25 @@ virtio_vdpa_virtq_enable(struct virtio_vdpa_priv *priv, int vq_idx)
 		return -EINVAL;
 	}
 
-	ret = virtio_vdpa_virtq_doorbell_relay_enable(priv, vq_idx);
-	if (ret) {
-		DRV_LOG(ERR, "%s virtq doorbell relay failed ret:%d",
-						priv->vdev->device->name, ret);
-		return ret;
+	if (priv->configured){
```
Comment (Ch3n60x) [inline drivers/vdpa/virtio/virtio_vdpa.c]: better add a space between ) and {

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 2]
```diff
@@ -583,11 +434,25 @@ virtio_vdpa_virtq_enable(struct virtio_vdpa_priv *priv, int vq_idx)
...
+		virtio_vdpa_lcore_id = rte_get_next_lcore(virtio_vdpa_lcore_id, 1, 1);
```
Comment (Ch3n60x) [inline drivers/vdpa/virtio/virtio_vdpa.c]: We get this lcore_id using rte_get_next_lcore every time. So do we need to define it as a global var? maybe temp var?

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 3]
```diff
@@ -583,11 +434,25 @@ virtio_vdpa_virtq_enable(struct virtio_vdpa_priv *priv, int vq_idx)
...
+		notify_work = rte_zmalloc(NULL, sizeof(*notify_work), 0);
+		if(!notify_work) {
```
Comment (Ch3n60x) [inline drivers/vdpa/virtio/virtio_vdpa.c]: better add a space after if

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 4]
```diff
@@ -583,11 +434,25 @@ virtio_vdpa_virtq_enable(struct virtio_vdpa_priv *priv, int vq_idx)
...
+			DRV_LOG(ERR, "%s vfid %d failed request memory", priv->vdev->device->name, priv->vf_id);
```
Comment (Ch3n60x) [inline drivers/vdpa/virtio/virtio_vdpa.c]: Maybe "failed to alloc notify work"?

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 5]
```diff
@@ -964,6 +829,24 @@ virtio_vdpa_dev_close_work(void *arg)
 	priv->dev_work_flag = VIRTIO_VDPA_DEV_CLOSE_WORK_DONE;
 	return ret;
 }
+static int
```
Comment (Ch3n60x) [inline drivers/vdpa/virtio/virtio_vdpa.c]: maybe add a blank line before this

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 6]
```diff
@@ -1233,6 +1117,23 @@ virtio_vdpa_dev_config(int vid)
 	}
 	priv->lm_status = VIRTIO_S_RUNNING;
 
+	virtio_vdpa_lcore_id = rte_get_next_lcore(virtio_vdpa_lcore_id, 1, 1);
+	notify_work = rte_zmalloc(NULL, sizeof(*notify_work), 0);
+	if(!notify_work) {
+		DRV_LOG(ERR, "%s vfid %d failed request memory", vdev->device->name, priv->vf_id);
```
Comment (Ch3n60x) [inline drivers/vdpa/virtio/virtio_vdpa.c]: similar comment as above

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 7]
```diff
@@ -1233,6 +1117,23 @@ virtio_vdpa_dev_config(int vid)
 	}
 	priv->lm_status = VIRTIO_S_RUNNING;
 
+	virtio_vdpa_lcore_id = rte_get_next_lcore(virtio_vdpa_lcore_id, 1, 1);
+	notify_work = rte_zmalloc(NULL, sizeof(*notify_work), 0);
+	if(!notify_work) {
```
Comment (Ch3n60x) [inline drivers/vdpa/virtio/virtio_vdpa.c]: space after if

File: drivers/vdpa/virtio/virtio_vdpa.c [hunk 8]
```diff
@@ -1233,6 +1117,23 @@ virtio_vdpa_dev_config(int vid)
...
+		rte_errno = rte_errno ? rte_errno : ENOMEM;
+		return -rte_errno;
+	}
+
+	notify_work->priv = priv;
+	notify_work->vq_idx = RTE_VHOST_QUEUE_ALL;
+	DRV_LOG(INFO, "%s vfid %d launch notifier work lcore:%d", vdev->device->name, priv->vf_id, virtio_vdpa_lcore_id);
+	ret = rte_eal_remote_launch(virtio_vdpa_dev_notifier_work, notify_work, virtio_vdpa_lcore_id);
+	if (ret) {
+		DRV_LOG(ERR, "%s vfid %d failed launch notifier work ret:%d lcore:%d", vdev->device->name, priv->vf_id, ret, virtio_vdpa_lcore_id);
```
Comment (Ch3n60x) [inline drivers/vdpa/virtio/virtio_vdpa.c]: line 1130 and 1133 seems too long. maybe using two lines?

**Comment by Ch3n60x** [comment]
Did some test. result shows before this commit, set_vring_state could take more than 0.01s for some queue. But after this change, all queues set_vring_state should be below 0.0001s.

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 1 above)
Comment (kailiangz1) [reply drivers/vdpa/virtio/virtio_vdpa.c]: done

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 2 above)
Comment (kailiangz1) [reply drivers/vdpa/virtio/virtio_vdpa.c]: we use global to prevent use same core next time

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 3 above)
Comment (kailiangz1) [reply drivers/vdpa/virtio/virtio_vdpa.c]: done

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 4 above)
Comment (kailiangz1) [reply drivers/vdpa/virtio/virtio_vdpa.c]: done

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 6 above)
Comment (kailiangz1) [reply drivers/vdpa/virtio/virtio_vdpa.c]: done

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 5 above)
Comment (kailiangz1) [reply drivers/vdpa/virtio/virtio_vdpa.c]: done

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 7 above)
Comment (kailiangz1) [reply drivers/vdpa/virtio/virtio_vdpa.c]: done

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 8 above)
Comment (kailiangz1) [reply drivers/vdpa/virtio/virtio_vdpa.c]: done

File: drivers/vdpa/virtio/virtio_vdpa.c (hunk 2 above)
Comment (Ch3n60x) [reply drivers/vdpa/virtio/virtio_vdpa.c]: ok

**Review by 2308524 — APPROVED** [review]

**Pull request merged**